src/main/java/net/neoforged/neoforge/event/ValidityEvent.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.event;
+
+import java.util.Optional;
+import net.minecraft.core.BlockPos;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.inventory.AbstractContainerMenu;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.entity.SignBlockEntity;
+import net.neoforged.neoforge.event.entity.player.PlayerEvent;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Event to check if a menu can stay open or if the player has moved too far away from it.
+ * <p>
+ * This is intended for mods that allow a player remote access to blocks and their menus,
+ * but it also can be used to forcibly kick them out of menus.
+ * <p>
+ * You can query the result of the original distance check with {@link #getOriginalValue()}
+ * and set your override with {@link #setResult(boolean)}.
+ * <p>
+ * This base class is never posted (but it can be subscribed to, as it is not abstract),
+ * only one of its subclasses:
+ * <ul>
+ * <li>{@link ValidityEvent.Menu} for all kinds of menus,
+ * <li>{@link ValidityEvent.ContainerMenu} when a AbstractContainerMenu is available for the event, or
+ * <li>{@link ValidityEvent.Sign} for the checks done by signs which don't use menus but a Screen.
+ * </ul>
+ */
+public class ValidityEvent extends PlayerEvent {
+    private final Level level;
+    private final BlockPos blockPos;
+    private final double distance;
+    private final boolean originalValue;
+    private Optional<Boolean> result = Optional.empty();
+
+    ValidityEvent(Player player, Level level, BlockPos blockPos, double distance) {
+        super(player);
+        this.level = level;
+        this.blockPos = blockPos;
+        this.distance = distance;
+        this.originalValue = player.canInteractWithBlock(blockPos, distance);
+    }
+
+    public Level getLevel() {
+        return level;
+    }
+
+    public BlockPos getBlockPos() {
+        return blockPos;
+    }
+
+    public double getDistance() {
+        return distance;
+    }
+
+    public boolean getOriginalValue() {
+        return originalValue;
+    }
+
+    public void setResult(boolean result) {
+        this.result = Optional.of(result);
+    }
+
+    public boolean getResult() {
+        return result.orElse(originalValue);
+    }
+
+    public static class Menu extends ValidityEvent {
```

Comment:
My idea is make this the default impl and let `ContainerMenu` and `Sign` be the only nested impls.

Reply:
I had it this way around at first, but that lacked a way of indicating that there is no menu open. So any mod trusting that a "*Menu*ValidityEvent" would only be fired when there's a menu set for the player would work fine...until someone opened a sign's menu and it'd crash horribly. That's why I moved the Menu variant down to be a sibling of Sign.

Reply:
You know what, you're right. But I would personally say that `ValidityEvent` isn't a good name. Potential rename?

Reply:
I'm not happy with that name either and hoped to get a good suggestion here ;).

Reply:
How about MenuRangeCheckEvent? As it only replaces the range check part of the validity check...

Reply:
@HenryLoenwind `MenuDistanceValidationEvent`?

Reply:
> > I had it this way around at first, but that lacked a way of indicating that there is no menu open. So any mod trusting that a "MenuValidityEvent" would only be fired when there's a menu set for the player would work fine...until someone opened a sign's ~menu~ _screen_ and it'd crash horribly. That's why I moved the Menu variant down to be a sibling of Sign.
> 
> Just don't let the subscriber assert anything they don't know from the event, if they want to assert anything they should do that themselves. The child event should only provide basic information that is not open to the subscriber's scope.

I can't prevent modders from assuming a menu is open when they subscribe to a "menu" event. That's why the event provides the information "hey this is not for a menu". When firing the event, the code knows that and can pass it on.

> The `Sign` child event is pretty much pointless as the `SignBlockEntity` can be retrieved by `Level#getBlockEntity` anyways.

But how would a subscriber know that? They could guess based on what's at the blockpos, but I prefer telling them explicitly. Also, when I already have the BE on hand, why have the subscriber look it up again? That just wastes CPU cycles and handing it over in the event is as close to free as possible.

> Do note that modded blocks may also use a menuless screen, I don't believe they should be forced to create their own variant of the event like you did for signs.

And that is a valid point. Changing the "Sign" sub-event to "Screen" and the parameter to a BE certainly makes sense.

This leaves Screens that are bound to a location that has no BE and no menu. Those don't exist in vanilla, and so I'd be reluctant to add a specific event for those. Too much guesswork about how that would be used for my liking, although I could make the "Blockbased" constructor public...

> The same applies to the `ContainerEntity` child event, if you simply don't inline the original check in the constructor, there's no point separating it from the `Entity` child event. I believe there should only be a `Block` child event providing only the `BlockPos` context, and a `Entity` child event providing only the `Entity` context,

The ContainerEntity event exists because the container that fires the event does not have access to the Entity (and vice versa). All I could provide is an AABB, which is mostly useless for finding the original Entity. For the same reason, the "Menu" event doesn't provide the menu/BE/etc. It is not available where it is fired, so the event subscriber's guess is as good as mine.

> the subscriber should retrieve specific context themselves and only make assertions based on their own checks (like `instanceof ContainerEntity`). Considering entity also has a valid `BlockPos`, and a position is needed anyways as the distance is in the base event,

The Entity has a pos, the ContainerEntity doesn't. That's why I removed Level and BlockPos from the base event. I could fill them in the Entity case (as I originally had it), but after pulling it down, that wasn't needed anymore. There's no advantage to putting stuff from entity in 2 additional fields when the entity itself is available.

> the `Block` child can optionally be pulled up to the top level, and let the `Entity` child be a separated entity specific implementation.

Making the top-level event abstract and opening up the Blockbased one instead is something I'd be fine with. Block-based and entity-based have diverged enough to make subscribing to the top-level event mostly pointless.

Reply:
PS: `ContainerEntity` extends `Container`, NOT `Entity`. Its name is wardsback.